documentation/src/main/asciidoc/reference/search-dsl-highlighting.asciidoc

```diff
@@ -0,0 +1,473 @@
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-dsl]
+----
+<1> Start the definition of a boundary scanner.
+<2> Pick a boundary scanner type -- a word scanner in this case.
+<3> Set an optional locale.
+<4> End the definition of a boundary scanner.
+<5> Set any other options.
+====
+
+.Setting the boundary scanner using lambda
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-lambda]
+----
+<1> Pass a lambda configurer to set up a boundary scanner.
+<2> Set any other options
+====
+
+Alternatively, a fast vector highlighter can use a character boundary scanner which relies on two other configurations --
+boundary characters and boundary max scan. When a character boundary scanner is used after a highlighted fragment is formed
+with highlighted text centred, the highlighter checks for the first occurrence of any configured boundary characters to the left and right
+of a currently created fragment. This lookup happens only for a maximum number of characters configured by the boundary max scan option.
+If no boundary characters are found, no additional text will be included besides the already highlighted phrase
+with surrounding text based on a fragment size option set for a highlighter.
+
+The default list of boundary characters includes `.,!? \t\n`.
+The default boundary max scan is equal to `20` characters.
+
+Character boundary scanner is a default option for the fast vector highlighters.
+
+.Setting the character boundary scanner
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/search/highlighting/HighlightingDslIT.java[tags=scanner-char]
+----
+Assuming that we have a text that contains multiple paragraphs separated with a new line (`\n`), we want to
+get the entire paragraph containing the highlighted phrase. To do so, boundary characters will be set to `\n`, and
+the max scan option will be based on the number of characters in paragraphs.
+
+<1> Start the definition of a boundary scanner.
+<2> Pick a boundary scanner type -- a character scanner.
+<3> Set a string containing boundary characters. The overloaded methods can accept a `String` or a `Character` array.
+<4> Set the max scan option.
+<5> End the definition of a boundary scanner.
+<6> Set any other options.
+====
+
+[NOTE]
+====
+This option is supported by the unified and fast vector highlighter types.
+====
+
+[[search-dsl-highlighting-highlighter-options-phrase-limit]]
+== Phrase limit
+Phrase limit allows to specify the maximum number of matching phrases in a document for highlighting.
+The highlighter will be going through the text and as soon as it performs the phrase limit number of highlights it'll stop
+leaving any further occurrences as not highlighted.
```

Comment:
Can you document the difference between this and the "number of fragments" above? You explained it in a comment, but not in the documentation:

> mmm, so the number of fragments is the number of items in the returned list. but this phrase limit - is how many phrases will get "tagged" so if we have a text with a word that is often repeated and we search for it - highlighter will wrap that word for `Phrase limit` times and all of them might be in the first fragment or it may span between all the fragments.
> 
> here's an example where we searched for a `bank`:
> 
> ```
> // note that there's a second "bank" occurrence, but since we only wanted one phrase highlighted it is ignored:
> contacts a police inspector and the manager of a nearby <em>bank</em>. With Watson, they hide in the bank vault and catch
> ```
> 
> the second bank is ignored because the phrase limit was set to 1.


Maybe something like this:

```suggestion
The highlighter will be going through the text and as soon as it reaches the maximum number of highlighted phrases it'll stop leaving any further occurrences as not highlighted.

[NOTE]
====
This limit is different from the <<search-dsl-highlighting-highlighter-options-fragment-size,maximum number of fragments>>:

1. Fragments are the strings returned by the highlight projections, while phrases are the sequences of highlighted (matching) terms in each fragment. A fragment may include multiple highlighted phrases, and a given phrase may appear in multiple fragments.
2. The phrase limit is about limiting highlighting of multiple occurrences of a *same phrase*, so with a phrase limit of `1` the phrase `shiny robot` will only be highlighted once, but there may be other highlighted phrases such as `great detective` (also highlighted only once).
==== 
```

Is this correct?

Reply:
Almost, I think there's a single "counter", so only `shiny robot` will get highlighted... I'll update the `2` point to:

2. The phrase limit is about limiting the highlighting of occurrences of matched phrases, be it multiple occurrences of the same phrase or a mix of different phrases.
For example, if we were to search for `fox` and `dog` occurrences in the sentence `The quick brown fox jumps over the lazy dog` and set the phrase limit to `1`, then we'll have only `fox` being highlighted since it was the first match in the text and the phrase limit was reached.